src/binary-encoding.adoc

```diff
@@ -0,0 +1,1600 @@
+C0
+----
+
+===== Example encoding of `{$10: 1, $11: 2}`
+[source]
+----
+┌──── An opcode in the range 0xC0-0xCF indicates a struct with symbol address field names
+│  ┌─── Field name: VarUInt 10 ($10)
+│  │        ┌─── Field name: VarUInt 11 ($11)
+│  │        │
+C6 15 51 01 17 51 02
+      └─┬─┘    └─┬─┘
+        1        2
+----
+
+===== Example encoding of `{$10: "variable length struct"}`
+[source]
+----
+ ┌───────────── Opcode `FC` indicates a variable length struct with symbol address field names
+ │  ┌────────── Length: VarUInt 25
+ │  │  ┌─────── Field name: VarUInt 10 ($10)
+ │  │  │  ┌──── Opcode `F8` indicates a variable length string
+ │  │  │  │  ┌─ VarUInt: 22 the string is 22 bytes long
+ │  │  │  │  │  v  a  r  i  a  b  l  e     l  e  n  g  t  h     s  t  r  u  c  t
+FC 33 15 F8 2D 76 61 72 69 61 62 6c 65 20 6c 65 6e 67 74 68 20 73 74 72 75 63 74
+               └─────────────────────────────┬─────────────────────────────────┘
+                                        UTF-8 bytes
+----
+
+[[structs_with_varsym_field_names]]
+==== Structs with `VarSym` field names
+
+NOTE: This encoding is very similar to <<structs_with_symbol_address_field_names, structs with symbol address
+field names>>, but allows writers to choose between representing each field name as a symbol address
+(for example: `$10`) or as inline UTF-8 bytes (for example: `"foo"`). This encoding is potentially less
+dense, but offers writers significant flexibility over whether and when field names are added to the
+symbol table.
+
+An opcode with a high nibble of `0xD_` indicates a struct with <<varsym, `VarSym`>> field names. The lower
+nibble of the opcode indicates how many bytes were used to encode all of its nested `(field name, value)`
+pairs.
+
+WARNING: Empty structs MUST be written using `0xC0`. `0xD0` is a reserved opcode.
+
+If the struct's encoded byte-length is too large to be encoded in a nibble, writers may use the `0xFD` opcode
+to write a variable-length struct with <<varsym, `VarSym`>> field names. The `0xFD` opcode is followed by a
+<<varuint, `VarUInt`>> that indicates the byte length.
+
+Each field in the struct is encoded as a  <<varsym, `VarSym`>> field name, followed by an
+opcode-prefixed value.
+
+===== Example encoding of `{"foo": 1, $11: 2}`
+[source]
+----
+┌─── Opcode with high nibble `D` indicates a struct with VarSym field names
+│┌── Length: 7
+││ ┌─ VarSym -3      ┌─ VarSym: 11 ($11)
+││ │   f  o  o       │
+D9 FD 66 6F 6F 51 01 17 91 02
+      └──┬───┘ └─┬─┘    └─┬─┘
+      3 UTF-8    1        2
+       bytes
+----
+
+// TODO: Demonstrate splicing macro values into the struct via VarSym escape code `0x00`.
+
+[[delimited_structs]]
+==== Delimited structs
+
+Opcode `0xF3` indicates the beginning of a delimited struct with `VarSym` field names,
+while opcode `0xF0` closes the most recently opened delimited container that has not yet been closed.
+
+NOTE: There is no delimited encoding for structs with symbol address field names.
```

Comment:
```suggestion
NOTE: There is no delimited encoding for structs that exclusively use symbol address field names.
```

Reply:
I've been using the phrase "structs with symbol address field names" to refer to the `VarUInt`-centric struct encoding. I'll try to clarify this with an extra clause.